build.rs

```diff
@@ -4,12 +4,69 @@ extern crate pkg_config;
 extern crate vcpkg;
 
 use std::env;
-use std::process::Command;
+#[allow(unused_imports)]
+use std::path::PathBuf;
+
+include!("./src/helper.rs");
 
 fn main() {
+    let include_dir = if env::var("MYSQLCLIENT_INCLUDE_DIR").is_ok() {
+        env::var("MYSQLCLIENT_INCLUDE_DIR").unwrap()
+    } else {
+        match mysql_config_variable("pkgincludedir") {
+            None => "".to_string(),
+            Some(var_value) => var_value,
+        }
+    };
+
+    let (version_id, is_mysql, is_mariadb) = get_libmysql_version_id(include_dir.clone());
+    if is_mysql {
+        if version_id >= 80010 {
+            println!("cargo::rustc-cfg=MySql_Version_AboveOrEqual_8_0_1");
+        }
+    } else if is_mariadb {
+        // MariaDB would Remove my_bool in future?
+        println!("cargo::rustc-cfg=mariadb_version_id=\"{}\"", version_id);
+    }
+
+    #[cfg(feature = "buildtime_bindgen")]
+    let mut use_rust_bindgen = true;
+    #[cfg(not(feature = "buildtime_bindgen"))]
+    let mut use_rust_bindgen = false;
+
+    if cfg!(target_os = "windows") {
+        use_rust_bindgen = false;
+    }
+
+    if env::var("USE_RUST_BINDGEN").is_ok() {
+        use_rust_bindgen = match env::var("USE_RUST_BINDGEN").unwrap().as_str() {
+            "true" => true,
+            "1" => true,
+            "" => true,
+            "false" => false,
+            "0" => false,
+            _ => false,
+        }
+    }
+
+    if use_rust_bindgen {
+        let lib_dir = if env::var("MYSQLCLIENT_LIB_DIR").is_ok() {
+            env::var("MYSQLCLIENT_LIB_DIR").unwrap()
+        } else {
+            match mysql_config_variable("pkglibdir") {
+                None => "".to_string(),
+                Some(var_value) => var_value,
+            }
+        };
+
+        if autogen_bindings(lib_dir, include_dir) == 0 {
+            return;
+        }
+    }
+
     if pkg_config::probe_library("mysqlclient").is_ok() {
         // pkg_config did everything for us
-        return
+        return;
     } else if try_vcpkg() {
         // vcpkg did everything for us
         return;
```

Comment:
For the `MYSQLCLIENT_LIB_DIR` variant I would expect the user to also provide a `MYSQLCLIENT_LIB_VERSION` variable for setting the library version. Otherwise we should give an explicit panic that requests the user to set this variable as well

For the `mysql_config_variable` variant: We can also query the library version from there. For the final variant that just says: Link these, we should likely also require the user to set `MYSQLCLIENT_LIB_VERSION`.

I feel that this might be much more robust than trying to parse c code to find a version somehow.